src/linalg.jl

```diff
@@ -224,87 +224,65 @@ end
 
 @inline dot(a::StaticVector, b::StaticVector) = _vecdot(same_size(a, b), a, b)
 @generated function _vecdot(::Size{S}, a::StaticArray, b::StaticArray) where {S}
-    if prod(S) == 0
-        return :(zero(promote_op(*, eltype(a), eltype(b))))
-    end
-
-    expr = :(conj(a[1]) * b[1])
-    for j = 2:prod(S)
-        expr = :($expr + conj(a[$j]) * b[$j])
-    end
-
     return quote
         @_inline_meta
-        @inbounds return $expr
+        s = zero(promote_op(*, eltype(a), eltype(b)))
```

Comment:
Pity... I was just deleting code which relies on inference...

Reply:
I can move this to only be used in the zero length case again.

Reply:
Yes, that would be good I think. Will the SIMD loop vectoriser still work if you peel off the first element in the loop, though? (Or maybe there is a better way?)

By the way, these were the one place where the `Union{}` eltype created errors (because of `zero(Union{})`.